l3kernel/l3str-convert.dtx

```diff
@@ -2567,6 +2580,102 @@
 \group_end:
 %    \end{macrocode}
 %
+% \subsection{PDF names and strings by expansion}
+%
+% \begin{macro}[EXP]{\str_convert_pdfname:n, \@@_convert_pdfname:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_first:w, \@@_convert_pdfname_loop:w}
+% \begin{macro}[EXP]{\@@_convert_pdfname_outer:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_inner:w}
+% \begin{macro}[EXP]
+%   {\@@_convert_pdfname_bytes:n, \@@_convert_pdfname_bytes_aux:n}
+% \begin{macro}[EXP]{\@@_convert_pdfname_bytes_aux:nnn}
+%   To convert to PDF names by expansion, we work purely on UTF-8 input. The
+%   first step is to make a string, after which we use a space-based mapping to
+%   divide up the input, then a simple token-by-token approach. In Unicode
+%   engines, we break down everything before one-byte codepoints, but for
+%   $8$-bit engines there is no need to worry. Actual escaping is covered
+%   by the same code as used in the non-expandable route.
+%    \begin{macrocode}
+\cs_new:Npn \str_convert_pdfname:n #1
+  {
+    \exp_args:Ne \tl_to_str:n
+      {
+         \exp_after:wN \@@_convert_pdfname:n \exp_after:wN
+           { \tl_to_str:n {#1} }
+      }
+  }
+\cs_new:Npx \@@_convert_pdfname:n #1
+  {
+    \exp_not:N \@@_convert_pdfname_first:w #1 ~
+      \exp_not:N \q_recursion_tail \c_space_tl
+      \exp_not:N \q_recursion_stop
+  }
```

Comment:
There shouldn't need to be a loop over space-delimited words and then over characters within a word.  I expect that applying `\__kernel_str_to_other_fast:n` within `e`-expansion is just as fast.  Then you could simply have a loop like
```
\cs_new:Npn \@@_convert_pdfname:n #1
  {
    \@@_convert_pdfname_loop:n #1 
    { ? \prg_break: }
    \prg_break_point:
  }
\cs_new:Npn \@@_convert_pdfname_loop:n #1
  {
    \use_none:n #1                               % This avoids a slow quark test
    \@@_convert_pdfname_one:N #1
    \@@_convert_pdfname_loop:n
  }
\bool_lazy_or:nnTF
  { \sys_if_engine_luatex_p: }
  { \sys_if_engine_xetex_p: }
  { ... }
  { \cs_new_eq:NN \@@_convert_pdfname_one:N \@@_escape_name_char:N }
```
Also since the `scan-quark` branch was merged, don't use `\q_recursion_tail` etc.

Reply:
I just thought today the clearest by far would be to use `\str_map_function:nN {#1} \@@_convert_pdfname_one:N`.  It should be of comparable speed to the current version, and would benefit from any speed improvement in `\str_map_function:nN`.